Content.Shared/_Moffstation/Cards/Systems/PlayingCardsSystem.Transfer.cs

```diff
@@ -0,0 +1,351 @@
+    /// transfer. The <paramref name="user"/> is also provided so that things like predicted audio can be played.
+    private delegate (Entity<PlayingCardComponent>? firstCard, int numTotalCards) CardSink(
+        IEnumerable<CardLike> cards,
+        EntityUid? user
+    );
+
+    /// This type represents a Card either as an entity with <see cref="PlayingCardComponent"/> or an unspawned card in
+    /// a <see cref="PlayingCardDeckComponent"/> during transfers.
+    private abstract record CardLike : ISealedInheritance
+    {
+        // Private constructor to seal inheritance.
+        private CardLike() { }
+
+        public sealed record Entity(Entity<PlayingCardComponent> Ent) : CardLike;
+
+        public sealed record Unspawned(PlayingCardInDeck Data) : CardLike;
+    }
+
+
+    /// <summary>
+    /// The implementation of all card movement into and out of stacks.
+    /// </summary>
+    /// <param name="source">The implementation of taking and returning cards from the source</param>
+    /// <param name="sourceCoords">Where the source is, conceptually. Movement animations and audio play from here</param>
+    /// <param name="sink">The implementation of adding cards to the sink</param>
+    /// <param name="sinkCoords">Where the sink is, conceptually. Movement animations and audio play from here</param>
+    /// <param name="range">Specification of which cards are taken. This is relative to <paramref name="source"/>.</param>
+    /// <param name="user">Who is causing the moving. This is used to make audio / pickup animations predictive. If this
+    /// is not specified, neither audio nor pickup animations will occur.</param>
+    private void TransferImpl(
+        CardSource source,
+        EntityCoordinates sourceCoords,
+        CardSink sink,
+        EntityCoordinates sinkCoords,
+        Range range,
+        EntityUid? user
+    )
+    {
+        var (firstCard, numTotalCards) = sink(source(range, user), user);
+
+        // TODO CENT If more than one card is moved, the pickup animation should look like a deck.
+        if (user is { } u && firstCard is { } fc)
+        {
+            _storage.PlayPickupAnimation(fc, sourceCoords, sinkCoords, 0, u);
+        }
+    }
+
+    private CardSource CardSourceFrom<TStack>(Entity<TStack> entity) where TStack : PlayingCardStackComponent =>
+        entity.Comp switch
+        {
+            PlayingCardDeckComponent deck => AsCardSource((entity, deck)),
+            PlayingCardHandComponent hand => AsCardSource((entity, hand)),
+            _ => throw new($"Unknown variant of {nameof(PlayingCardStackComponent)}"),
+        };
+
+    private CardSource AsCardSource(Entity<PlayingCardDeckComponent> entity)
+    {
+        return Impl;
+
+        // Local method to enable `yield`.
+        IEnumerable<CardLike> Impl(Range range, EntityUid? user)
+        {
+            var (inRange, outOfRange) = Split(entity.Comp.Cards, range);
+            try
```

Comment:
`try`/`finally` in here allows the stacks to perform guaranteed maintenance operations after the sourcing/sinking is done. This means that no matter what happens, audio, dirtying, degeneracy checks, etc. _will_ happen.